File: apps/social/src/lib/ext-agent-slash-commands.ts
import type {
  ExtAgentCommandCatalog,
  ExtAgentCommandDescriptor,
  MmxMediaKind,
  RunMmxMediaCommandParams,
  RunMmxMediaCommandResult,
} from "@envoymesh/api";

/** True when the composer value is an in-progress slash token (no args yet). */
export function isExtAgentSlashSuggestInput(value: string): boolean {
  return /^\/\S*$/.test(value);
}

/** Prefix used to filter catalog rows (`"/mod"` → `/mod…`). */
export function extAgentSlashPrefix(value: string): string {
  if (!isExtAgentSlashSuggestInput(value)) return "";
  return value.toLowerCase();
}

export function filterExtAgentSlashCommands(
  commands: ExtAgentCommandDescriptor[],
  value: string,
): ExtAgentCommandDescriptor[] {
  const prefix = extAgentSlashPrefix(value);
  if (!prefix) return [];
  return commands.filter((cmd) => cmd.slash.toLowerCase().startsWith(prefix));
}

/** `/help` (optionally with trailing args) — Envoy-handled locally. */
export function isExtAgentHelpCommand(text: string): boolean {
  const trimmed = text.trim();
  if (!trimmed.startsWith("/")) return false;
  const [cmd] = trimmed.slice(1).split(/\s+/);
  return (cmd ?? "").toLowerCase() === "help";
}

export type ExtAgentModelSlashAction =
  | { type: "show" }
  | { type: "list" }
  | { type: "default" }
  | { type: "set"; model: string };

/** Parse `/model` [show|list|default|<id>] — null if not a /model command. */
export function parseExtAgentModelCommand(text: string): ExtAgentModelSlashAction | null {
  const trimmed = text.trim();
  if (!trimmed.startsWith("/")) return null;
  const parts = trimmed.slice(1).split(/\s+/).filter(Boolean);
  const cmd = parts[0]?.toLowerCase() ?? "";
  if (cmd !== "model") return null;
  const rest = parts.slice(1).join(" ").trim();
  if (!rest || rest.toLowerCase() === "show") return { type: "show" };
  if (rest.toLowerCase() === "list") return { type: "list" };
  if (rest.toLowerCase() === "default") return { type: "default" };
  return { type: "set", model: rest };
}

const MMX_MEDIA_KINDS = new Set<string>([
  "image",
  "video",
  "speech",
  "music",
  "vision",
  "search",
  "quota",
  "auth",
]);

export type ParsedMmxMediaCommand =
  | { ok: true; params: RunMmxMediaCommandParams }
  | { ok: false; error: string }
  | null;

/**
 * Parse MiniMax media slash commands.
 * `/mmx-auth` maps to kind `auth`.
 * Returns null when the text is not an MMX media command.
 */
export function parseMmxMediaCommand(text: string): ParsedMmxMediaCommand {
  const trimmed = text.trim();
  if (!trimmed.startsWith("/")) return null;
  const parts = trimmed.slice(1).split(/\s+/).filter(Boolean);
  if (parts.length === 0) return null;
  let cmd = (parts[0] ?? "").toLowerCase();
  if (cmd === "mmx-auth") cmd = "auth";
  if (!MMX_MEDIA_KINDS.has(cmd)) return null;
  const kind = cmd as MmxMediaKind;
  const rest = parts.slice(1);

  if (kind === "quota" || kind === "auth") {
    return { ok: true, params: { kind } };
  }
  if (kind === "vision") {
    const target = rest[0]?.trim();
    if (!target) {
      return { ok: false, error: "Usage: /vision <path-or-url> [question]" };
    }
    const question = rest.slice(1).join(" ").trim();
    return {
      ok: true,
      params: {
        kind,
        target,
        ...(question ? { prompt: question } : {}),
      },
    };
  }
  const prompt = rest.join(" ").trim();
  if (!prompt) {
    const usage =
      kind === "speech"
        ? "Usage: /speech <text>"
        : kind === "search"
          ? "Usage: /search <query>"
          : `Usage: /${kind} <prompt>`;
    return { ok: false, error: usage };
  }
  return { ok: true, params: { kind, prompt } };
}

export function formatMmxMediaResult(result: RunMmxMediaCommandResult): string {
  if (!result.ok) {
    return `MiniMax /${result.kind} failed: ${result.error ?? "unknown error"}`;
  }
  const lines: string[] = [`MiniMax /${result.kind}`];
  if (result.path) lines.push(`Saved: ${result.path}`);
  if (result.text?.trim()) lines.push(result.text.trim());
  return lines.join("\n");
}

/** `/model` with optional arg prefix — for model-id autocomplete. */
export function extAgentModelSuggestPrefix(value: string): string | null {
  const match = value.match(/^\/model(?:\s+(.*))?$/i);
  if (!match) return null;
  // Still typing `/model` without trailing space → slash command menu, not models.
  if (!/\s/.test(value)) return null;
  return (match[1] ?? "").toLowerCase();
}

export function filterExtAgentModels(
  models: Array<{ id: string; label?: string }>,
  value: string,
): Array<{ id: string; label?: string }> {
  const prefix = extAgentModelSuggestPrefix(value);
  if (prefix === null) return [];
  if (!prefix) return models;
  return models.filter(
    (m) =>
      m.id.toLowerCase().startsWith(prefix) ||
      (m.label?.toLowerCase().startsWith(prefix) ?? false),
  );
}

export function formatExtAgentSlashHelp(catalog: ExtAgentCommandCatalog): string {
  const lines: string[] = [`${catalog.agentName} slash commands:`];
  if (catalog.commands.length === 0) {
    lines.push("(none)");
  } else {
    for (const cmd of catalog.commands) {
      const args = cmd.argsHint ? ` ${cmd.argsHint}` : "";
      lines.push(`${cmd.slash}${args} — ${cmd.summary}`);
    }
  }
  if (catalog.supportsSessionModel) {
    const current = catalog.sessionModel ?? catalog.defaultModel ?? "(default)";
    lines.push("");
    lines.push(`Current model: ${current}`);
  }
  if (catalog.limitations?.length) {
    lines.push("");
    lines.push("Notes:");
    for (const note of catalog.limitations) {
      lines.push(`• ${note}`);
    }
  }
  return lines.join("\n");
}

export function formatExtAgentModelShow(catalog: ExtAgentCommandCatalog): string {
  const current = catalog.sessionModel
    ? `${catalog.sessionModel} (session override)`
    : `${catalog.defaultModel ?? "(default)"} (default)`;
  return `${catalog.agentName} model: ${current}`;
}

export function formatExtAgentModelList(catalog: ExtAgentCommandCatalog): string {
  const models = catalog.models ?? [];
  if (models.length === 0) {
    return `${catalog.agentName}: no model list available. Try /model <id> if you know the id.`;
  }
  const lines = [`${catalog.agentName} models:`];
  for (const m of models) {
    lines.push(m.label ? `• ${m.id} — ${m.label}` : `• ${m.id}`);
  }
  return lines.join("\n");
}

File: apps/envoygo/lib/ext_agent/ext_agent_slash_commands.dart
/// Ext Agent slash helpers for EnvoyGo (mirrors Social `ext-agent-slash-commands.ts`).

bool isExtAgentSlashSuggestInput(String value) {
  return RegExp(r'^/\S*$').hasMatch(value);
}

bool isExtAgentHelpCommand(String text) {
  final trimmed = text.trim();
  if (!trimmed.startsWith('/')) return false;
  final parts = trimmed.substring(1).split(RegExp(r'\s+'));
  final cmd = parts.isEmpty ? '' : parts.first;
  return cmd.toLowerCase() == 'help';
}

class ExtAgentModelSlashAction {
  final String type; // show | list | default | set
  final String? model;
  const ExtAgentModelSlashAction.show() : type = 'show', model = null;
  const ExtAgentModelSlashAction.list() : type = 'list', model = null;
  const ExtAgentModelSlashAction.default_() : type = 'default', model = null;
  const ExtAgentModelSlashAction.set(this.model) : type = 'set';
}

ExtAgentModelSlashAction? parseExtAgentModelCommand(String text) {
  final trimmed = text.trim();
  if (!trimmed.startsWith('/')) return null;
  final parts = trimmed.substring(1).split(RegExp(r'\s+')).where((p) => p.isNotEmpty).toList();
  if (parts.isEmpty || parts.first.toLowerCase() != 'model') return null;
  final rest = parts.skip(1).join(' ').trim();
  if (rest.isEmpty || rest.toLowerCase() == 'show') {
    return const ExtAgentModelSlashAction.show();
  }
  if (rest.toLowerCase() == 'list') return const ExtAgentModelSlashAction.list();
  if (rest.toLowerCase() == 'default') {
    return const ExtAgentModelSlashAction.default_();
  }
  return ExtAgentModelSlashAction.set(rest);
}

const _mmxMediaKinds = {
  'image',
  'video',
  'speech',
  'music',
  'vision',
  'search',
  'quota',
  'auth',
};

/// Parse result for MiniMax media slash commands.
class ParsedMmxMediaCommand {
  final bool ok;
  final Map<String, String>? params;
  final String? error;
  const ParsedMmxMediaCommand.ok(this.params)
      : ok = true,
        error = null;
  const ParsedMmxMediaCommand.err(this.error)
      : ok = false,
        params = null;
}

/// Parse MiniMax media slash. `/mmx-auth` maps to kind `auth`.
/// Returns null when the text is not an MMX media command.
ParsedMmxMediaCommand? parseMmxMediaCommand(String text) {
  final trimmed = text.trim();
  if (!trimmed.startsWith('/')) return null;
  final parts =
      trimmed.substring(1).split(RegExp(r'\s+')).where((p) => p.isNotEmpty).toList();
  if (parts.isEmpty) return null;
  var cmd = parts.first.toLowerCase();
  if (cmd == 'mmx-auth') cmd = 'auth';
  if (!_mmxMediaKinds.contains(cmd)) return null;
  final rest = parts.skip(1).toList();

  if (cmd == 'quota' || cmd == 'auth') {
    return ParsedMmxMediaCommand.ok({'kind': cmd});
  }
  if (cmd == 'vision') {
    if (rest.isEmpty) {
      return const ParsedMmxMediaCommand.err(
        'Usage: /vision <path-or-url> [question]',
      );
    }
    final target = rest.first.trim();
    final question = rest.skip(1).join(' ').trim();
    return ParsedMmxMediaCommand.ok({
      'kind': cmd,
      'target': target,
      if (question.isNotEmpty) 'prompt': question,
    });
  }
  final prompt = rest.join(' ').trim();
  if (prompt.isEmpty) {
    final usage = cmd == 'speech'
        ? 'Usage: /speech <text>'
        : cmd == 'search'
            ? 'Usage: /search <query>'
            : 'Usage: /$cmd <prompt>';
    return ParsedMmxMediaCommand.err(usage);
  }
  return ParsedMmxMediaCommand.ok({'kind': cmd, 'prompt': prompt});
}

String formatMmxMediaResult(Map<String, dynamic> result) {
  final kind = result['kind']?.toString() ?? 'media';
  final ok = result['ok'] == true;
  if (!ok) {
    return 'MiniMax /$kind failed: ${result['error'] ?? 'unknown error'}';
  }
  final lines = <String>['MiniMax /$kind'];
  final path = result['path']?.toString().trim();
  if (path != null && path.isNotEmpty) lines.add('Saved: $path');
  final text = result['text']?.toString().trim();
  if (text != null && text.isNotEmpty) lines.add(text);
  return lines.join('\n');
}

String? extAgentModelSuggestPrefix(String value) {
  final match = RegExp(r'^/model(?:\s+(.*))?$', caseSensitive: false).firstMatch(value);
  if (match == null) return null;
  if (!value.contains(RegExp(r'\s'))) return null;
  return (match.group(1) ?? '').toLowerCase();
}

List<Map<String, dynamic>> filterExtAgentSlashCommands(
  List<Map<String, dynamic>> commands,
  String value,
) {
  if (!isExtAgentSlashSuggestInput(value)) return const [];
  final prefix = value.toLowerCase();
  return commands.where((c) {
    final slash = (c['slash'] as String?)?.toLowerCase() ?? '';
    return slash.startsWith(prefix);
  }).toList();
}

List<Map<String, dynamic>> filterExtAgentModels(
  List<Map<String, dynamic>> models,
  String value,
) {
  final prefix = extAgentModelSuggestPrefix(value);
  if (prefix == null) return const [];
  if (prefix.isEmpty) return models;
  return models.where((m) {
    final id = (m['id'] as String?)?.toLowerCase() ?? '';
    final label = (m['label'] as String?)?.toLowerCase() ?? '';
    return id.startsWith(prefix) || label.startsWith(prefix);
  }).toList();
}

String formatExtAgentSlashHelp(Map<String, dynamic> catalog) {
  final name = (catalog['agentName'] as String?)?.trim().isNotEmpty == true
      ? catalog['agentName'] as String
      : (catalog['agentId'] as String? ?? 'Ext Agent');
  final lines = <String>['$name slash commands:'];
  final commands = (catalog['commands'] as List?) ?? const [];
  if (commands.isEmpty) {
    lines.add('(none)');
  } else {
    for (final raw in commands) {
      if (raw is! Map) continue;
      final slash = raw['slash']?.toString() ?? '';
      final args = (raw['argsHint'] as String?)?.trim();
      final summary = raw['summary']?.toString() ?? '';
      lines.add(args == null || args.isEmpty ? '$slash — $summary' : '$slash $args — $summary');
    }
  }
  if (catalog['supportsSessionModel'] == true) {
    final current = (catalog['sessionModel'] as String?) ??
        (catalog['defaultModel'] as String?) ??
        '(default)';
    lines.add('');
    lines.add('Current model: $current');
  }
  final limitations = (catalog['limitations'] as List?) ?? const [];
  if (limitations.isNotEmpty) {
    lines.add('');
    lines.add('Notes:');
    for (final note in limitations) {
      lines.add('• $note');
    }
  }
  return lines.join('\n');
}

String formatExtAgentModelShow(Map<String, dynamic> catalog) {
  final name = (catalog['agentName'] as String?) ?? 'Ext Agent';
  final session = catalog['sessionModel'] as String?;
  final current = session != null
      ? '$session (session override)'
      : '${catalog['defaultModel'] ?? '(default)'} (default)';
  return '$name model: $current';
}

String formatExtAgentModelList(Map<String, dynamic> catalog) {
  final name = (catalog['agentName'] as String?) ?? 'Ext Agent';
  final models = (catalog['models'] as List?) ?? const [];
  if (models.isEmpty) {
    return '$name: no model list available. Try /model <id> if you know the id.';
  }
  final lines = <String>['$name models:'];
  for (final raw in models) {
    if (raw is! Map) continue;
    final id = raw['id']?.toString() ?? '';
    final label = raw['label']?.toString();
    lines.add(label == null || label.isEmpty ? '• $id' : '• $id — $label');
  }
  return lines.join('\n');
}
